File: scripts/dependencies_annotate_dependencies.py
import pandas as pd
import os

def main():

    ###  SNAKEMAKE I/O ###
    raw_crispr_data    = snakemake.input['crispr_gene_dependency_chronos']
    lines_info         = snakemake.input['sample_info']
    ccle_raw_reads     = snakemake.input['raw_expected_counts']
    where_to_save      = snakemake.output['model_candidates']
    

    # Prepare a dict. to convert DepMap IDs to stripped cell line name
    lines  = pd.read_csv(lines_info,
                         sep=',',
                         usecols=['DepMap_ID', 'stripped_cell_line_name', 'lineage'],
                         index_col=0
                         )
    
    human_readable = lines['stripped_cell_line_name'].to_dict()
    lines_lineage  = lines['lineage'].to_dict()

    raw_crispr_data = pd.read_csv(raw_crispr_data, sep=',', index_col=0)

    # Drop CRISPR profiled lines not present in RNA-Seq assay
    rna_seq_lines         = pd.read_csv(ccle_raw_reads, usecols=[0])
    rna_seq_lines.columns = ['profiled_lines']

    raw_crispr_data = raw_crispr_data[raw_crispr_data.index.isin(rna_seq_lines['profiled_lines'])]
    
    del rna_seq_lines

    # Transpose for better performance. Genes are rows now
    raw_crispr_data = raw_crispr_data.T

    # Get rid of genes with NaN values, usually due to defective processing/contamination
    filtered_crispr_data = raw_crispr_data.dropna(axis=0)

      # Drop engineered cell lines, If present
    is_engineered = lines.loc[lines['lineage'].str.contains('engineered'), 'stripped_cell_line_name']
    filtered_crispr_data = filtered_crispr_data.loc[:,~filtered_crispr_data.columns.isin(is_engineered)]

    # Get index as columns to avoid R calling it 'X'
    filtered_crispr_data.reset_index(drop=False, inplace=True)
    filtered_crispr_data.rename({'index':'Gene'}, axis=1, inplace=True)

    # CCLE nomenclature is HUGO (ENTREZ). Keep HUGO ONLY
    filtered_crispr_data['variance'] = filtered_crispr_data.var(axis=1)
    filtered_crispr_data['Gene']     = [x.split(' ')[0] for x in filtered_crispr_data['Gene'].values.tolist()]

    ## Resolve duplicates keeping most variable gene
    filtered_crispr_data_uniques = filtered_crispr_data.sort_values('variance', ascending=False).drop_duplicates('Gene', keep='first').sort_index()	
    filtered_crispr_data_uniques.drop('variance', axis=1, inplace=True)

    # melt the dataset 
    melted_crispr = filtered_crispr_data_uniques.melt(id_vars=['Gene'], var_name='cell_line', value_name='probability')

    # Filter out genes with less than 5 cell lines dependant on them
    melted_crispr            = melted_crispr.groupby('Gene').filter(lambda x: len(x[x['probability'] >= 0.95]) >= 5)

    # annotate the lineage and stripped name
    melted_crispr['lineage']            = melted_crispr['cell_line'].map(lines_lineage)
    melted_crispr['stripped_line_name'] = melted_crispr['cell_line'].map(human_readable)

    # TODO: we could add lineages before saving the models in the future
    # for now, It's pancancer
    melted_by_gene = melted_crispr.groupby('Gene')

    for gene_model in melted_by_gene.groups.items():

        model_data = melted_crispr.loc[gene_model[1]]
        gene_name  = model_data['Gene'].unique()[0]
        
        if not os.path.exists(where_to_save):
            os.mkdir(where_to_save)

        model_data.to_csv(f"{where_to_save}/{gene_name}.csv",sep=',', index=False, header=True)

    

if __name__ == "__main__":
    main()

File: workflow/scripts/dependencies_annotate_dependencies.R
library("tidyverse")

###  SNAKEMAKE I/O ###
raw_crispr_data = snakemake@input[["crispr_gene_dependency_chronos"]]
lines_info = snakemake@input[["sample_info"]]
ccle_raw_reads = snakemake@input[["expression_matrix"]]
where_to_save = snakemake@output[["model_candidates"]]

ccle_counts <- readRDS(ccle_raw_reads)
crispr <- data.table::fread(raw_crispr_data)
annot <-  data.table::fread(lines_info)

## keep lines whose primary disease is cancer AND with available expr.
annot_cancer <- annot %>%
    filter(
        !primary_disease %in% c("Unknown", "Non-Cancerous") 
    )

## drop lineages with too few CCLs
all_lineages <- table(annot_cancer$lineage)
lineages_to_keep <- names(all_lineages[all_lineages >= 5])

## get the names of the CCLs whose lineages are OK AND with RNA-seq
lines_to_keep <- annot_cancer %>%
    filter(
        lineage %in% lineages_to_keep
    ) %>%
    pull(DepMap_ID)

lines_to_keep <- intersect(lines_to_keep, colnames(ccle_counts))

## Filter crispr data
crispr_filtered <- crispr %>%
    filter(
        ModelID %in% lines_to_keep
    )

crispr_lines <- crispr_filtered$ModelID
crispr_filtered$ModelID <- NULL

# CCLE nomenclature is HUGO (ENTREZ). Keep HUGO ONLY
colnames(crispr_filtered) <- stringr::str_remove_all(string = colnames(crispr_filtered),
                                                     pattern = " \\(.*$"
                                                     )

## Resolve duplicates keeping most variable gene
crispr_filtered <- t(crispr_filtered)
colnames(crispr_filtered) <- crispr_lines

vars <- apply(crispr_filtered, 1, var)
crispr_filtered <- cbind(crispr_filtered, vars)
crispr_filtered <- crispr_filtered[order(crispr_filtered[, "vars"],decreasing=TRUE),]

crispr_filtered_unique <- crispr_filtered[!duplicated(rownames(crispr_filtered)), ]
crispr_filtered_unique <- crispr_filtered_unique[, colnames(crispr_filtered_unique) != "vars"]

## Melt the dataset
crispr_probabilities <- crispr_filtered_unique %>%
    as_tibble(rownames = "Gene") %>%
    pivot_longer(
        cols = starts_with("ACH-"),
        names_to = "cell_line",
        values_to = "probability"
        )


# Filter out genes with less than 5 cell lines dependant on them
crispr_enough_power <- crispr_probabilities %>%
    group_by(Gene) %>%
    mutate(dependent_lines = sum(probability >= 0.5)) %>%
    filter(
        dependent_lines >= 5
    )

# annotate the lineage and stripped name
crispr_enough_power_annotated <- crispr_enough_power %>%
    left_join(
        y = annot_cancer[, c("DepMap_ID",
                             "lineage",
                             "stripped_cell_line_name",
                             "original_lineage"
                             )
                         ],
        by = c("cell_line" = "DepMap_ID")
    )


if(!dir.exists(file.path(where_to_save))){
    dir.create(where_to_save)}

by(crispr_enough_power_annotated,
   crispr_enough_power_annotated$Gene,
   FUN=function(i) write.csv(i, paste0(where_to_save, "/", i$Gene[1], ".csv"), , row.names=FALSE))
